### ISSUE TITLE
[Coupon] - Coupon amount not deducted from order total if coupon added before product added

### ISSUE THREAD

**Issue by jostnes** [woothemes/woocommerce#32596]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Found this from testing p91TBi-6ZF-p2 
During order creation, if a coupon is added before a product is added, the coupon amount is not reflected in the total amount.

Screen recording:

https://user-images.githubusercontent.com/17252150/163086850-7897ece8-6db8-469f-8dad-32d0486edf56.mov

### Expected behavior

The coupon amount should be reflected in the total amount even if it's added before a product is added.

### Actual behavior

The coupon amount is not reflected in the order total.

### Steps to reproduce

1. Create an order on the web
2. Add a coupon before adding the product
3. Complete the order - See that the coupon amount is not applied to the order
4. Create another order 
5. Add the product before adding the coupon - See that the coupon amount is applied correctly

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://jddwootest.wpcomstaging.com
Site address (URL): https://jddwootest.wpcomstaging.com
WC Version: 6.4.0
REST API Version: ✔ 6.4.0
WC Blocks Version: ✔ 7.2.1
Action Scheduler Version: ✔ 3.4.0
WC Admin Version: ✔ 3.3.2
Log Directory Writable: ✔
WP Version: 5.9.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_GB
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 7.4.28
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 7.82.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.3.34-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.4.0
WC Database Prefix: wp_
Total Database Size: 3.87MB
Database Data Size: 2.35MB
Database Index Size: 1.52MB
wp_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.05MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.28MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 1.02MB + Index: 0.09MB + Engine InnoDB
wp_posts: Data: 0.09MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

amp_validated_url: 1
attachment: 5
jetpack_migration: 2
nav_menu_item: 1
page: 7
post: 1
product: 8
revision: 5
shop_coupon: 5
shop_order: 29
shop_order_refund: 2
wp_global_styles: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (11) ###

Akismet Anti-Spam: by Automattic – 4.2.2
AMP: by AMP project contributors – 2.2.4
Crowdsignal Forms: by Automattic – 1.6.0
WordPress.com Editing Toolkit: by Automattic – 3.30687
Gutenberg: by Gutenberg Team – 12.9.0
Jetpack: by Automattic – 10.9-a.1
Layout Grid: by Automattic – 1.8.1
Page Optimize: by Automattic – 0.5.2
Crowdsignal Polls & Ratings: by Automattic
Inc. – 3.0.7

WooCommerce Payments: by Automattic – 3.9.3
WooCommerce: by Automattic – 6.4.0

### Inactive Plugins (1) ###

Classic Editor: by WordPress Contributors – 1.6.2

### Dropin Plugins (2) ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Must Use Plugins (1) ###

WP.com Site Helper: by  –

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #13 - /shop/
Basket: #14 - /basket/
Checkout: #15 - /checkout/
My account: #16 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Two
Version: 1.0.1-wpcom (update to version 1.1 is available)
Author URL: https://en-gb.wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build
personally we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –

### WooCommerce Payments ###

Version: 3.9.3
Connected to WPCOM: Yes
Blog ID: 199997875
Account ID: -

### Action Scheduler ###

Canceled: 1
Oldest: 2022-03-21 16:04:08 +0800
Newest: 2022-03-21 16:04:08 +0800

Complete: 62
Oldest: 2022-03-14 16:34:26 +0800
Newest: 2022-04-13 10:09:15 +0800

Pending: 1
Oldest: 2022-04-13 11:14:51 +0800
Newest: 2022-04-13 11:14:51 +0800

### Status report information ###

Generated at: 2022-04-13 10:25:35 +08:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by anandrajaram21** [comment]
Hey, I'd like to work on this issue, can I be assigned to it? I was able to reproduce this issue as mentioned in the issue description. A fix for this includes adding an extra conditional inside the maybe_add_order_item function in class-wc-ajax.php that checks if there are any coupons have been applied and then run the recalculate_coupons() function to calculate the new order total

**Comment by pmcpinto** [comment]
Routing this to engineering for the next step: the report looks actionable enough for implementation or documentation work.
